charmhelpers/core/hookenv.py

```diff
@@ -1088,22 +1097,31 @@ def function_tag():
     return os.environ.get('JUJU_FUNCTION_TAG') or action_tag()
 
 
-def status_set(workload_state, message):
+def status_set(workload_state, message, application_status=False):
     """Set the workload state with a message
 
     Use status-set to set the workload state with a message which is visible
     to the user via juju status. If the status-set command is not found then
     assume this is juju < 1.23 and juju-log the message instead.
 
-    workload_state -- valid juju workload state.
-    message        -- status update message
+    workload_state     -- valid juju workload state. str or WL_STATES
+    message            -- status update message
+    application_status -- Whether this is an application state set
     """
-    valid_states = ['maintenance', 'blocked', 'waiting', 'active']
-    if workload_state not in valid_states:
+    # Extract the value if workload_state is an Enum
+    try:
+        workload_state = workload_state.value
+    except AttributeError:
+        pass
+    workload_state = workload_state.lower()
+    if workload_state not in [s.lower() for s in WL_STATES.__members__.keys()]:
         raise ValueError(
             '{!r} is not a valid workload state'.format(workload_state)
         )
-    cmd = ['status-set', workload_state, message]
+    cmd = ['status-set']
+    if application_status:
+        cmd.append('--application')
+    cmd.extend([workload_state, message])
     try:
         ret = subprocess.call(cmd)
```

Comment:
We should be prepared for the hook to fail. `status-set --application` will error when it's called by a non-leader.

Perhaps we should call `is-leader` early and raise an error directly?

Reply:
I agree and I wish the original author (which unfortunately git blame seems to suggest was me) had done that but making that change now would be a change in behaviour on a function that has been around for 5 years. I don't want to inadvertently break any code that is expected a subprocess error when running on a non-leader